From: Marcelo Tosatti <mtosatti@redhat.com>
Date: Tue, 12 Jan 2010 21:33:53 -0500
Subject: [kvm] fix kvmclock-adjust-offset ioctl to match upstream
Message-id: <20100112214153.649798476@amt.cnet>
Patchwork-id: 22466
O-Subject: [RHEL 6 KVM 04/20] KVM: fix kvmclock-adjust-offset ioctl to match
	upstream
Bugzilla: 554524
RH-Acked-by: Avi Kivity <avi@redhat.com>
RH-Acked-by: Gleb Natapov <gleb@redhat.com>
RH-Acked-by: Juan Quintela <quintela@redhat.com>

The ioctl number and kvm_clock_data structure imported in the Fedora
kernel do not match upstream. Fix it.

Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
Bugzilla: 554524

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/Documentation/kvm/api.txt b/Documentation/kvm/api.txt
index 2353c4b..db3a706 100644
--- a/Documentation/kvm/api.txt
+++ b/Documentation/kvm/api.txt
@@ -607,7 +607,8 @@ such as migration.
 
 struct kvm_clock_data {
 	__u64 clock;  /* kvmclock current value */
-	__u64 pad[2];
+	__u32 flags;
+	__u32 pad[9];
 };
 
 4.28 KVM_SET_CLOCK
@@ -624,7 +625,8 @@ such as migration.
 
 struct kvm_clock_data {
 	__u64 clock;  /* kvmclock current value */
-	__u64 pad[2];
+	__u32 flags;
+	__u32 pad[9];
 };
 
 5. The kvm_run structure
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index c1dc4a1..74eed09 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -2593,6 +2593,10 @@ long kvm_arch_vm_ioctl(struct file *filp,
 		if (copy_from_user(&user_ns, argp, sizeof(user_ns)))
 			goto out;
 
+		r = -EINVAL;
+		if (user_ns.flags)
+			goto out;
+
 		r = 0;
 		ktime_get_ts(&now);
 		now_ns = timespec_to_ns(&now);
@@ -2608,10 +2612,13 @@ long kvm_arch_vm_ioctl(struct file *filp,
 		ktime_get_ts(&now);
 		now_ns = timespec_to_ns(&now);
 		user_ns.clock = kvm->arch.kvmclock_offset + now_ns;
+		user_ns.flags = 0;
 
+		r = -EFAULT;
 		if (copy_to_user(argp, &user_ns, sizeof(user_ns)))
-			r =  -EFAULT;
+			goto out;
 
+		r = 0;
 		break;
 	}
 
diff --git a/include/linux/kvm.h b/include/linux/kvm.h
index 87890b3..d572437 100644
--- a/include/linux/kvm.h
+++ b/include/linux/kvm.h
@@ -439,7 +439,7 @@ struct kvm_ioeventfd {
 #endif
 #define KVM_CAP_IOEVENTFD 36
 #define KVM_CAP_SET_IDENTITY_MAP_ADDR 37
-#define KVM_CAP_ADJUST_CLOCK 38
+#define KVM_CAP_ADJUST_CLOCK 39
 
 #ifdef KVM_CAP_IRQ_ROUTING
 
@@ -504,7 +504,8 @@ struct kvm_irqfd {
 
 struct kvm_clock_data {
 	__u64 clock;
-	__u64 pad[2];
+	__u32 flags;
+	__u32 pad[9];
 };
 
 /*
